### PR TITLE
Bind nonlinear binary-code profiles to explicit codes

### LINKED ISSUE #1785: [Coding theory] Add exact nonlinear binary code and constant-weight distance-profile operations
## Audit context

Audited at `14957005b6bc691c9142818949cb7f05d082e713`.

`conjecture-probes-v1/constant-weight-code-a23-6-10-2992` accepts any explicit 2992-word binary code of length 23, constant weight 6, and minimum Hamming distance at least 10. The verifier does not trust the public archive file; it replays every word and distance predicate.

Current `src/jacobian/domains/code_theory/operations.py` only owns generator-matrix linear codes (`minimum_distance`, `weight_distribution`). An arbitrary constant-weight code need not be linear and has no generator matrix. Treating its 2992 represented words as a linear code is mathematically wrong; benchmark-local pairwise loops are reusable domain logic.

### `BinaryWord`

A fixed-length bit vector with a compact canonical representation and exact Hamming weight.

### `ExplicitBinaryCode`

A bounded finite set of distinct `BinaryWord` values sharing one length. Word order is not mathematical; duplicates are invalid or surfaced before set construction.

### `ConstantWeightBinaryCode`

An `ExplicitBinaryCode` whose words all have one exact weight `w`. This narrow value makes mixed-weight states unrepresentable.

### `code.binary.explicit.profile.compute`

Return:

- length and cardinality;
- duplicate/malformed rows before strict construction;
- complete weight distribution;
- exact minimum and maximum pairwise Hamming distance;
- distance distribution;
- deterministic pairs attaining each extremum; and
- complete pair-count accounting.

For large codes, emit a compact distance histogram plus first/extremal witnesses, not millions of pair rows. A separately named materialized pair ledger may have a much smaller bound.

### `code.binary.constant_weight.profile.compute`

Input a `ConstantWeightBinaryCode`; return `n`, `w`, `M`, minimum distance, distance histogram, intersection-size histogram under

```text
d_H(x,y) = 2(w-|supp(x) intersection supp(y)|),
```

and extremal word/support pairs. Replay both bitwise and support-intersection formulas on bounded fixtures.

### `code.binary.word_distance.compute`

Direct exact leaf for two equal-length words, returning differing coordinates, distance, both weights, and support intersection. This supports composition without invoking a whole-code profile.

### `code.binary.explicit.to_set_system.compute`

Map words to their support subsets on coordinate labels, with an exact word-to-block bijection. Reuse the finite set-system/incidence owner rather than creating a coding-specific incidence representation.

### `code.binary.constant_weight.distance_graph.compute`

For supplied distance predicate/value `d`, return the existing finite graph whose vertices are codeword IDs and whose edges bind exact word pairs at that distance. This is a direct transform, not clique/code search.

## Semantics and bounds

- Coordinate permutation and coherent word relabelling preserve scalar profiles.
- Codeword order is irrelevant.
- Complementing every word maps weight `w` to `n-w` and preserves pairwise distance.
- Duplicate words are not two codewords in an ordinary set code.
- Runtime/output failure cannot become a larger minimum distance.

Preflight-bound length, code cardinality, total bits, pair count `M(M-1)/2`, bitset work, histogram bins, materialized witnesses, and output bytes. The compact profile should support the 2992-word fixture without returning all pair rows.

## Benchmarks

Cover:

1. tiny linear and nonlinear codes;
2. constant- and mixed-weight families;
3. duplicates and length mismatch;
4. exact distance/support identities;
5. coordinate permutation and complement metamorphisms;
6. compact versus materialized profile agreement on small codes;
7. the complete `A(23,6,10)>=2992` witness as an integration fixture;
8. one bit mutation lowering distance or changing weight; and
9. exact-at-bound and over-bound requests.

## Acceptance criteria

- [ ] Explicit nonlinear binary codes and constant-weight codes have closed bounded values distinct from generator-matrix linear codes.
- [ ] Profiles completely account for every unordered pair and return exact minima/histograms.
- [ ] The 2992-word benchmark can be replayed through a compact bounded operation.
- [ ] Support-set and distance-graph transforms reuse existing set/graph values.
- [ ] No code-construction search, `A(n,d,w)` optimizer, archive/database integration, generic verifier framework, or theorem wrapper is introduced.

## Non-goals

- Finding a 2992-word code.
- Proving optimality of 2992.
- Replacing existing linear-code operations.
- Trusting an external code file without replay.

### PR DESCRIPTION
**Pull request by morluto**
Closes #1785

## Why

The nonlinear-code profile operations accepted a whole code as an unbound tuple of bit words. That lost the ambient coordinate axis for an empty code, allowed producer and consumer contracts to drift, and did not make the profile's exact pairwise source sufficiently explicit for downstream composition.

## What changed

- Introduce canonical `ExplicitBinaryCode(length, codewords)` as the whole-code value. It preserves the coordinate axis, canonicalizes word order, and admits the two natural zero-coordinate codes: the empty family and the singleton empty word.
- Bind explicit and constant-weight profile/source/support operations to that value; keep pairwise distance computation as a native integer-bitset kernel.
- Return exact pair totals, histograms, deterministic extrema, and source-indexed witnesses, with producer/result replay rejecting forged metadata, sources, histograms, axes, supports, and witnesses.
- Add derived admission for materialized bitsets, pair count, two-pass scan work, retained source, and serialized result size. Empty and singleton profiles correctly omit nonexistent witnesses from that output budget.
- Retain the full A(23,6,10) 2,992-word regression fixture (length 23, minimum distance 6, weight 10) and its complete distance histogram.

The implementation deliberately does not introduce a distance graph. These operations expose reusable exact code-profile values; a graph would be a separate mathematical postcondition.

## Validation

- `make check` — 3,551 passed (one existing unrelated syntax warning)
- `make test-math TESTS=tests/math/code_nonlinear/test_code_nonlinear.py` — 46 passed

Suggested review order: `values.py` and `_budget.py` define the canonical value and finite envelope; `_models.py` and `_operations.py` bind profile semantics and replay; focused tests cover the standard code, output boundaries, zero-coordinate degeneracies, and mutation rejection.

### CODE REVIEW

**Comment by morluto** [comment]
@codex

File: src/jacobian/math/code_nonlinear/values.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            )
+        },
+    )
+
+    length: Annotated[
+        int,
+        Field(strict=True, ge=0, le=MAX_EXPLICIT_CODE_LENGTH),
+    ]
+    codewords: tuple[BinaryWord, ...] = Field(
+        default=(),
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/code_nonlinear/values.py:57]: Root cause: ExplicitBinaryCode checked length*cardinality only in a mode="after" validator, so Pydantic converted and strict-validated every supplied bit and word before any bound ran; the raw containers were unbounded.

Fix (src/jacobian/math/code_nonlinear/values.py): added a mode="before" model validator on ExplicitBinaryCode that rejects raw shapes before nested materialization — more than MAX_EXPLICIT_CODE_BITS (524288) codewords ("word container bound") or cumulative raw entries exceeding the 524288-bit source bound with early exit ("bit source bound"). Also made the limits schema-visible with Field max_length=MAX_EXPLICIT_CODE_BITS on the codewords tuple (maxItems) and on the canonical BinaryWord alias; verified this stacks cleanly with the existing per-field constraints in WordDistanceRequest.

Tests (tests/math/code_nonlinear/test_code_nonlinear.py): test_millions_of_empty_words_rejected_before_nested_conversion, test_aggregate_bit_overflow_rejected_before_nested_conversion, test_enormous_single_word_rejected_before_nested_conversion — each plants an invalid element that nested conversion would surface as a different error, so the asserted bound message proves rejection precedes materialization; test_exact_aggregate_boundary_passes_prevalidation pins acceptance at exactly MAX_EXPLICIT_CODE_BITS entries. Validation: pytest tests/math/code_nonlinear 51 passed; ruff and mypy clean.

File: src/jacobian/math/code_nonlinear/_models.py [hunk 2]
```diff
@@ -50,190 +81,371 @@ def require_valid_weight(self) -> Self:
...
+                "0 or 1. The exact result, including every differing coordinate, "
+                "must fit the 4194304-byte result bound."
+            )
+        },
+    )
 
-    word1: tuple[int, ...] = Field(min_length=1, max_length=MAX_LENGTH)
-    word2: tuple[int, ...] = Field(min_length=1, max_length=MAX_LENGTH)
+    word1: BinaryWord = Field(min_length=1, max_length=MAX_EXPLICIT_CODE_LENGTH)
+    word2: BinaryWord = Field(min_length=1, max_length=MAX_EXPLICIT_CODE_LENGTH)
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/code_nonlinear/_models.py:142]: Root cause: this PR changed the word-distance wire contract (accepted length 16 -> 524288, plus strict bit and coordinate constraints) but the declaration still used _op's default version="1" in src/jacobian/math/code_nonlinear/_tools.py, so version-keyed clients would treat the incompatible schema as the old contract.

Fix: code.binary.word_distance.compute now passes version="2", matching the other changed operations in this PR.

Test: added TestWordDistance::test_contract_version_tracks_the_wire_shape_change pinning the declared version to "2". Validation: make test-catalog 29 passed (catalog snapshot/invariant behavior) and make test-dispatch 13 passed.

File: src/jacobian/math/code_nonlinear/_models.py [hunk 3]
```diff
@@ -1,46 +1,77 @@
...
 
 
 class ConstantWeightRequest(StrictModel):
-    """Generate all constant-weight binary words of given length and weight."""
+    """Generate all binary words of one bounded length and weight."""
 
-    length: int = Field(ge=1, le=MAX_LENGTH)
-    weight: int = Field(ge=0)
+    length: Annotated[int, Field(strict=True, ge=0, le=MAX_GENERATED_LENGTH)]
+    weight: Annotated[int, Field(strict=True, ge=0, le=MAX_GENERATED_LENGTH)]
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/code_nonlinear/_models.py]: Root cause: `ConstantWeightRequest`/`ConstantWeightResult` imposed a flat `le=16` on length and weight, rejecting small-work cases like {length: 64, weight: 0} that the previous length-64 contract admitted. Admission now derives the envelope from the mathematics: `require_constant_weight_admission` in `src/jacobian/math/code_nonlinear/_budget.py` bounds `length * binomial(length, weight)` (which simultaneously bounds combinations work, intermediates, and the exact generated code) at the 524288-entry canonical explicit-code source bound, plus an explicit result wire check against the 4 MiB result bound; fields now use the coordinate-axis maximum. Changed `_budget.py`, `_models.py`, and the operation description in `_tools.py`. Added boundary tests (64/0, 64/1, axis-length 524288/0, admitted 17/8 with replay) and the central-binomial rejection (64/32) in `tests/math/code_nonlinear/test_code_nonlinear.py`. Validated with `make check` (3917 passed), `make test-math TESTS="tests/math/code_nonlinear"` (53 passed), and `make test-catalog` (34 passed).

File: src/jacobian/math/code_nonlinear/_budget.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    # Fixed object keys, braces, commas, and scalar punctuation fit well
+    # below this 64-byte allowance.
+    return 64 + words_bytes + 3 * (_digits(length) + _digits(cardinality))
+
+
+def require_constant_weight_admission(length: int, weight: int) -> int:
+    """Admit the complete weight-``weight`` generation before enumerating."""
+    if not 0 <= weight <= length:
+        raise ValueError("weight cannot exceed length")
+    cardinality = comb(length, weight)
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/code_nonlinear/_budget.py]: Confirmed at HEAD: require_constant_weight_admission(524288, 262144) spent ~2s in math.comb computing the 524,279-bit central coefficient, then the f-string interpolation of the ~524k-digit entries count raised Python integer-to-decimal conversion error ("Exceeds the limit (4300 digits)") instead of the intended entry-bound ValueError.

Fixed by replacing the full comb() call with a thresholded multiplicative computation (_binomial_within_entry_budget): it iterates C(length - w + k, k) step by step, checks length * partial against MAX_GENERATED_CODE_ENTRIES before every multiplication (including the initial length * 1), and raises the identical "entry bound" error at the first crossing. After symmetry reduction (w := min(w, length - w)) each step multiplies by at least two, so rejection happens within ~log2(budget) steps (~0ms measured), all reported counts stay small enough to interpolate, and the accepted/rejected domain is unchanged because the partial entries sequence is monotonically nondecreasing toward the exact binomial. Accepted requests still return the exact comb(length, weight).

Tests: test_constant_weight_admission_is_exact_within_entry_bound (known-answer exactness incl. symmetry), test_central_weight_rejection_stops_before_the_full_binomial (asserts the digit-limit host error is gone and the crossing message reports 524288 coordinates * 262145 words), test_constant_weight_request_rejects_at_the_immediate_entry_boundary ((101,2) admitted vs (102,2) rejected plus the max-length central-weight request). Validation: pytest tests/math/code_nonlinear -n 4 --timeout=120 (59 passed), make test-catalog (34 passed), make test-dispatch (24 passed), ruff and mypy clean.

[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/code_nonlinear/_budget.py]: Confirmed with your exact request: `length=524288, weight=262144` spent ~3.4s computing the full ~524k-bit central binomial and then failed inside the f-string on Python's integer-to-decimal digit limit instead of raising the intended entry-bound error. The fix replaces the whole-coefficient computation with a thresholded multiplicative binomial (`_binomial_within_entry_budget`) that checks `length * cardinality` after each exact step and aborts at the first crossing, so rejection now takes microseconds and reports small, interpolable counts; an exactness sweep verifies admitted cardinalities still equal `math.comb`. Regression tests cover the maximum-length central weight, immediate-boundary cases like (102, 2), and exactness against `math.comb` for admitted pairs; the code_nonlinear lane passes (59 passed) and `make check` is green (3923 passed) on tip 545051e11.

File: src/jacobian/math/code_nonlinear/_models.py [hunk 5]
```diff
@@ -1,239 +1,448 @@
...
+    word1: BinaryWord = Field(min_length=1, max_length=MAX_EXPLICIT_CODE_LENGTH)
+    word2: BinaryWord = Field(min_length=1, max_length=MAX_EXPLICIT_CODE_LENGTH)
 
     @model_validator(mode="after")
     def require_valid_words(self) -> Self:
         if len(self.word1) != len(self.word2):
             raise ValueError("words must have equal length")
-        if any(b not in (0, 1) for b in self.word1 + self.word2):
-            raise ValueError("words must be binary (0 or 1)")
+        require_word_distance_output_bound(len(self.word1))
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/code_nonlinear/_models.py]: Reproduced first: two identical 524,288-bit zero words were rejected with the length-only estimate of 5,657,085 bytes against the 4 MiB limit, even though their canonical result is 2,097,262 bytes. `require_word_distance_output_bound()` now takes both words and charges the exact wire size of the actual differing coordinates instead of the full coordinate axis, so that request is admitted (request validation plus execution in about 0.24 s). The all-different worst case still computes exactly 5,657,085 bytes and is rejected unchanged, so no envelope was weakened. Added known-answer, canonical-bound invariant, adversarial, and coordinate-digit boundary tests in TestWordDistance.

File: src/jacobian/math/code_nonlinear/values.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+MAX_EXPLICIT_CODE_LENGTH = MAX_EXPLICIT_CODE_BITS
+
+BinaryBit = Annotated[int, Field(strict=True, ge=0, le=1)]
+type BinaryWord = Annotated[
+    tuple[BinaryBit, ...],
+    Field(max_length=MAX_EXPLICIT_CODE_LENGTH),
+]
+
+
+class ExplicitBinaryCode(StrictModel):
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/code_nonlinear/values.py:27]: Fixed in f9a2915ef. The nonlinear-code request and result models now use ExplicitBinaryCode as the canonical source, and the operation declarations execute through the source-bound value and profile admission. Empty and degenerate codes are supported, schemas publish code: {length, codewords}, and the old duplicated top-level codewords contract and validators were removed. Added regression coverage for the canonical JSON path and empty/singleton profiles.

File: src/jacobian/math/code_nonlinear/_operations.py [hunk 7]
```diff
@@ -1,198 +1,308 @@
...
+        coordinate_axis=tuple(range(code.length)),
+        supports=tuple(
+            tuple(i for i, bit in enumerate(word) if bit) for word in code.codewords
+        ),
     )
+
+
+def to_set_system(code: ExplicitBinaryCode) -> ToSetSystemResult:
+    """Native support conversion for one canonical explicit binary code."""
+    return compute_to_set_system(ToSetSystemRequest(code=code))
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/code_nonlinear/_operations.py]: Fixed in f9a2915ef. The native jacobian.math.code_nonlinear.to_set_system now accepts ExplicitBinaryCode and calls the typed support projection directly, without constructing a transport request or applying the MCP result-size bound. The MCP adapter remains separately admitted with its retained-source/output bound.

File: src/jacobian/math/code_nonlinear/values.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/code_nonlinear/values.py:27]: Follow-up: ae8b830a4 restores the source-bound profile/result contracts and exact witness replay that were present before the merge conflict, including empty/singleton semantics and producer-consumer closure. The focused nonlinear-code suite is now 64 passing tests.

File: src/jacobian/math/code_nonlinear/_operations.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (morluto) [reply src/jacobian/math/code_nonlinear/_operations.py]: Follow-up: ae8b830a4 also changes the native helper to construct ToSetSystemResult directly. Only the MCP request validator applies require_set_system_output_bound; native ExplicitBinaryCode callers are not routed through the wire admission.
